ddtrace/contrib/internal/openai/utils.py

```diff
@@ -162,6 +162,136 @@ async def _extract_token_chunk(self, chunk):
+                )
+            self._dd_span.finish()
+
+    def __next__(self):
+        try:
+            chunk = self.__wrapped__.__next__()
+            _loop_handler(self._dd_span, chunk, self._streamed_chunks)
+            return chunk
+        except StopIteration:
+            _process_finished_stream(
+                self._dd_integration, self._dd_span, self._kwargs, self._streamed_chunks, self._is_completion
+            )
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_span.finish()
+            raise
+
+
+class TracedOpenAIAsyncResponseStream(BaseTracedOpenAIStream):
+    async def __aenter__(self):
+        await self.__wrapped__.__aenter__()
+        return self
+
+    async def __aexit__(self, exc_type, exc_val, exc_tb):
+        await self.__wrapped__.__aexit__(exc_type, exc_val, exc_tb)
+
+    async def __aiter__(self):
+        exception_raised = False
+        try:
+            async for chunk in self.__wrapped__:
+                if chunk.type == "response.completed":
+                    # Set response attributes
+                    if hasattr(chunk.response, "id"):
+                        self._dd_span.set_tag("openai.response.id", chunk.response.id)
+                    if hasattr(chunk.response, "created_at"):
+                        self._dd_span.set_tag_str("openai.response.created_at", str(chunk.response.created_at))
+                    if hasattr(chunk.response, "model"):
+                        self._dd_span.set_tag("openai.response.model", chunk.response.model)
+                    handle_response_tools(chunk.response, self._dd_span)
+                    self._dd_integration.record_usage(self._dd_span, chunk.response.usage)
+                    self._dd_span.finish()
+                yield chunk
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            exception_raised = True
+            raise
+        finally:
+            if not exception_raised:
+                _process_finished_stream(
+                    self._dd_integration, self._dd_span, self._kwargs, self._streamed_chunks, self._is_completion
+                )
+            self._dd_span.finish()
+
+    async def __anext__(self):
+        try:
+            chunk = await self.__wrapped__.__anext__()
+            _loop_handler(self._dd_span, chunk, self._streamed_chunks)
+            return chunk
+        except StopAsyncIteration:
+            _process_finished_stream(
+                self._dd_integration, self._dd_span, self._kwargs, self._streamed_chunks, self._is_completion
+            )
+            self._dd_span.finish()
+            raise
+        except Exception:
+            self._dd_span.set_exc_info(*sys.exc_info())
+            self._dd_span.finish()
+            raise
+
+
+def handle_response_tools(resp, span):
```

Comment:
I think that instead of introducing a new helper method for responses, we can reuse `_tag_streamed_response()` or add this logic into the latter.